contracts/SwapAndBridge.sol

```diff
@@ -0,0 +1,158 @@
+     *     - flags: Overwritten by this contract.
+     * @param minExpectedInputTokenAmount Minimum amount of `dstToken` to receive after swap and to submit to Across
+     * as deposit.inputAmount.
+     * @param outputToken Token to receive from Across deposit on destination chain.
+     * @param outputAmount of `outputToken` to receive from Across deposit on
+     * destination chain.
+     */
+    function swap1InchAndBridge(
+        IAggregationExecutor aggregationExecutor,
+        I1InchAggregationRouterV5.SwapDescription memory swapDescription,
+        uint256 minExpectedInputTokenAmount,
+        address outputToken,
+        uint256 outputAmount,
+        address depositor,
+        address recipient,
+        uint256 destinationChainid,
+        address exclusiveRelayer,
+        uint32 quoteTimestamp,
+        uint32 fillDeadline,
+        uint32 exclusivityDeadline,
+        bytes memory message
+    ) external nonReentrant {
+        // @dev: Don't let caller use WETH, or better, hardcode the swap and acrossInput tokens to be USDC.e/USDC
+        // because 1Inch has special rules to handle WETH<-->ETH.
+
+        // Pull tokens from caller into this contract.
+        address swapToken = address(swapDescription.srcToken);
+        uint256 swapTokenAmount = swapDescription.amount;
+        IERC20(swapToken).transferFrom(msg.sender, address(this), swapTokenAmount);
+
+        // Craft 1Inch swap arguments to swap `swapToken` for `acrossInputToken` using this contract as the recipient
+        // of the swap.
+        swapDescription.srcReceiver = payable(address(this));
+        swapDescription.dstReceiver = payable(address(this));
+        swapDescription.flags = 0; // TODO: Change this if we want to support ETH swaps and/or partial fills (i.e.
+        // swapTokenAmount != spentAmount). For now, set to 0 for simplicity.
+
+        IERC20(swapDescription.dstToken).approve(address(oneInchRouter), swapTokenAmount);
+
+        uint256 balanceBefore = IERC20(swapDescription.dstToken).balanceOf(address(this));
+        // @dev: Example swap I used for comparison:
+        // https://optimistic.etherscan.io/tx/0x8a4e77ee1a62e94b42b21e849bcdabd60d43ac49191cd2878f6b47f395f26abc
+        (uint256 returnAmount, uint256 spentAmount) = oneInchRouter.swap(
+            aggregationExecutor,
+            swapDescription,
+            new bytes(0), // TODO: No IERC20Permit.permit needed since we're sending an approval?
+            new bytes(0) // TODO: We don't want to execute any data on swaps but I'm not sure how this is used.
+        );
+
+        // Sanity check that 1inch doesn't partial fill.
+        require(spentAmount == swapTokenAmount, "spent amount != swap amount");
+        // Sanity check that received amount from swap is enough to submit Across deposit with.
+        require(returnAmount >= minExpectedInputTokenAmount, "min expected input amount");
```

Comment:
Should we check the actual balance rather than depending on the 1inch output?

Reply:
3b57ed15e5bc0d851b0277caff304b4f09578191